relayer/chains/wasm/provider.go

```diff
@@ -575,36 +578,101 @@ func (p *Provider) fetchBlockMessages(ctx context.Context, heightInfo *types.Hei
+				}
+				if retryCount >= types.RPCMaxRetryAttempts {
+					p.logger.Error("fetchBlockMessages failed", zap.Uint8("attempt", retryCount), zap.Error(err))
+					errorChan <- err
+					break
+				}
+				p.logger.Error("failed to fetch block messages", zap.Uint64("start_height", localSearchParam.StartHeight), zap.Uint64("end_height", localSearchParam.EndHeight), zap.Error(err))
+				time.Sleep(delay)
 			}
-			if res.TotalCount > perPage {
-				for i := 2; i <= int(res.TotalCount/perPage)+1; i++ {
-					searchParam.Page = &i
-					resNext, err := p.client.TxSearch(ctx, searchParam)
+
+			if localMessages.TotalCount > perPage {
+				for i := 2; i <= int(localMessages.TotalCount/perPage)+1; i++ {
+					p.logger.Info("fetching block messages", zap.Uint64("start_height", localSearchParam.StartHeight), zap.Uint64("end_height", localSearchParam.EndHeight), zap.Int("page", i))
+					localSearchParam.Page = &i
+					resNext, err := p.client.TxSearch(ctx, localSearchParam)
 					if err != nil {
-						errorChan <- err
-						return
+						var attempts uint8
+						for attempts < types.RPCMaxRetryAttempts {
+							p.logger.Error("failed to fetch block messages with page", zap.Uint64("start_height", localSearchParam.StartHeight), zap.Uint64("end_height", localSearchParam.EndHeight), zap.Int("page", i), zap.Error(err))
+							attempts++
+							delay := time.Duration(math.Pow(2, float64(attempts))) * types.BaseRPCRetryDelay
+							if delay > types.MaxRPCRetryDelay {
+								delay = types.MaxRPCRetryDelay
+							}
+							if attempts >= types.RPCMaxRetryAttempts {
+								p.logger.Error("fetchBlockMessages failed", zap.Uint8("attempt", attempts), zap.Duration("retrying_in", delay), zap.Uint64("start_height", localSearchParam.StartHeight), zap.Uint64("end_height", localSearchParam.EndHeight))
+								errorChan <- err
+								break
+							}
+							p.logger.Warn("fetchBlockMessages failed, retrying...", zap.Uint8("attempt", attempts), zap.Duration("retrying_in", delay))
+							time.Sleep(delay)
+						}
 					}
-					res.Txs = append(res.Txs, resNext.Txs...)
+					localMessages.Txs = append(localMessages.Txs, resNext.Txs...)
 				}
 			}
-			messagesChan <- res
-		}(&wg, searchParam, messagesChan, errorChan)
+			messagesChan <- localMessages
+		}(event)
+	}
+
+	go func() {
+		wg.Wait()
+		close(messagesChan)
+		close(errorChan)
+	}()
+
+	var errors []error
+	for {
 		select {
-		case msgs := <-messagesChan:
-			messages.Txs = append(messages.Txs, msgs.Txs...)
-			messages.TotalCount += msgs.TotalCount
-		case err := <-errorChan:
-			p.logger.Error("failed to fetch block messages", zap.Error(err))
+		case msgs, ok := <-messagesChan:
+			if !ok {
+				messagesChan = nil
+			} else {
+				messages.Txs = append(messages.Txs, msgs.Txs...)
+				messages.TotalCount += msgs.TotalCount
+			}
+		case err, ok := <-errorChan:
+			if !ok {
+				errorChan = nil
+			} else {
+				errors = append(errors, err)
+			}
 		}
+		if messagesChan == nil && errorChan == nil {
```

Comment:
What happens we already ahve errors and we don;t get anything in the messages chan ?? Will it happen ?